include/z64transition.h

```diff
@@ -1,85 +1,103 @@
-} TransitionInstanceType;
+    /*  0 */ TRANS_TYPE_WIPE,
+    /*  1 */ TRANS_TYPE_TRIFORCE,
+    /*  2 */ TRANS_TYPE_FADE_BLACK,
+    /*  3 */ TRANS_TYPE_FADE_WHITE,
+    /*  4 */ TRANS_TYPE_FADE_BLACK_FAST,
+    /*  5 */ TRANS_TYPE_FADE_WHITE_FAST,
+    /*  6 */ TRANS_TYPE_FADE_BLACK_SLOW,
+    /*  7 */ TRANS_TYPE_FADE_WHITE_SLOW,
+    /*  8 */ TRANS_TYPE_WIPE_FAST,
+    /*  9 */ TRANS_TYPE_FILL_WHITE2,
+    /* 10 */ TRANS_TYPE_FILL_WHITE,
+    /* 11 */ TRANS_TYPE_INSTANT,
+    /* 12 */ TRANS_TYPE_FILL_BROWN,
+    /* 13 */ TRANS_TYPE_FADE_WHITE_CS_DELAYED,
+    /* 14 */ TRANS_TYPE_SANDSTORM_PERSIST,
+    /* 15 */ TRANS_TYPE_SANDSTORM_END,
+    /* 16 */ TRANS_TYPE_CS_BLACK_FILL,
+    /* 17 */ TRANS_TYPE_FADE_WHITE_INSTANT,
+    /* 18 */ TRANS_TYPE_FADE_GREEN,
+    /* 19 */ TRANS_TYPE_FADE_BLUE,
+    // transition types 20 - 31 are unused
+    // transition types 32 - 55 are constructed using the TRANS_TYPE_CIRCLE macro
+    /* 56 */ TRANS_TYPE_MAX = 56
+} TransitionType;
 
-typedef struct {
-    /* 0x000 */ Color_RGBA8_u32 color;
-    /* 0x004 */ Color_RGBA8_u32 unkColor;
-    /* 0x008 */ u8 direction;
-    /* 0x009 */ u8 frame;
-    /* 0x00A */ u8 isDone;
-    /* 0x00C */ u16 texX;
-    /* 0x00E */ u16 texY;
-    /* 0x010 */ u16 normal;
-    /* 0x018 */ Mtx projection;
-    /* 0x058 */ Mtx lookAt;
-    /* 0x098 */ Mtx modelView[2][3];
-} TransitionWipe; // size = 0x218
+#define TRANS_NEXT_TYPE_DEFAULT 0xFF // when `nextTransitionType` is set to default, the type will be taken from the entrance table for the ending transition
 
-#define TRANS_INSTANCE_TYPE_FADE_FLASH 3
+typedef enum {
+    /* 0 */ TCA_NORMAL,
+    /* 1 */ TCA_WAVE,
+    /* 2 */ TCA_RIPPLE,
+    /* 3 */ TCA_STARBURST
+} TransitionCircleAppearance;
 
-typedef struct {
-    /* 0x000 */ u8 type;
-    /* 0x001 */ u8 isDone;
-    /* 0x002 */ u8 direction;
-    /* 0x004 */ Color_RGBA8_u32 color;
-    /* 0x008 */ u16 timer;
-} TransitionFade; // size = 0xC
+typedef enum {
+    /* 0 */ TCC_BLACK,
+    /* 1 */ TCC_WHITE,
+    /* 2 */ TCC_GRAY,
+    /* 3 */ TCC_SPECIAL // color varies depending on appearance. unused and appears broken
+} TransitionCircleColor;
 
-typedef struct {
-    /* 0x000 */ Color_RGBA8_u32 color;
-    /* 0x004 */ Color_RGBA8_u32 unkColor;
-    /* 0x008 */ s32 texX;
-    /* 0x00C */ s32 texY;
-    /* 0x010 */ s32 speed;
-    /* 0x014 */ u8 direction;
-    /* 0x015 */ u8 colorType;
-    /* 0x016 */ u8 speedType;
-    /* 0x017 */ u8 appearanceType;
-    /* 0x018 */ u8 isDone;
-    /* 0x019 */ u8 frame;
-    /* 0x01A */ u16 normal;
-    /* 0x020 */ Mtx projection;
-    /* 0x060 */ Mtx lookAt;
-    /* 0x0A0 */ void* texture;
-    /* 0x0A8 */ Mtx modelView[2][3];
-} TransitionCircle; // size = 0x228;
+typedef enum {
+    /* 0 */ TCS_FAST,
+    /* 1 */ TCS_SLOW
+} TransitionCircleSpeed;
+
+#define TC_SET_PARAMS (1 << 7)
+
+#define TRANS_TYPE_CIRCLE(appearance, color, speed) ((1 << 5) | ((color & 3) << 3) | ((appearance & 3) << 1) | (speed & 1))
```

Comment:
Should probably go in `z64transition_instances` since they are `circle` specific.

Reply:
I'm conflicted

The enums like `TransitionCircleSpeed` would belong in `z64transition_instances.h` since the transition circle implementation uses them

But `TRANS_TYPE_CIRCLE` is a z_play/transition system thing (it constructs a transition type), so I wouldn't put it there?

But also it would be weird to split them up? (though it's probably the best thing to do idk)

Reply:
Yeah idk `TRANS_TYPE_CIRCLE` is weird because it both fits in the 32-55 range of `TransitionType`, which is z_play-specific, and is also relevant in `TransitionCircle_SetType` which unpacks the bitfield

Plus there's the `TC_SET_PARAMS` hack with `TransitionCircle_SetType` being called twice, the first time for params, the second time for direction

Reply:
True, that is weird. Personally I would probably say leave `TRANS_TYPE_CIRCLE` and probably `TC_SET_PARAMS` but move the enums to `z64transition_instances`, but I'll leave it up to you.